src/planner/binder/query_node/bind_macro_select_node.cpp

```diff
@@ -0,0 +1,165 @@
+		// if expr is a parameter, replace it with its argument
+		auto &colref = (ColumnRefExpression &)*expr;
+		bool bind_macro_parameter = false;
+		if (colref.IsQualified()) {
+			bind_macro_parameter = colref.GetTableName() == MacroBinding::MACRO_NAME;
+		} else {
+			bind_macro_parameter = macro_binding->HasMatchingBinding(colref.GetColumnName());
+		}
+		if (bind_macro_parameter) {
+			D_ASSERT(macro_binding->HasMatchingBinding(colref.GetColumnName()));
+			expr = macro_binding->ParamToArg(colref);
+		}
+		return;
+	}
+	case ExpressionClass::SUBQUERY: {
+		// replacing parameters within a subquery is slightly different
+		auto &sq = ((SubqueryExpression &)*expr).subquery;
+		ParsedExpressionIterator::EnumerateQueryNodeChildren(*sq->node, [&](unique_ptr<ParsedExpression> &child) {
+			ReplaceMacroSelectParametersRecursive(child, macro_binding);
+		});
+		break;
+	}
+	default: // fall through
+		break;
+	}
+	// unfold child expressions
+	ParsedExpressionIterator::EnumerateChildren(*expr, [&](unique_ptr<ParsedExpression> &child) {
+		ReplaceMacroSelectParametersRecursive(child, macro_binding);
+	});
+}
+
+unique_ptr<QueryNode> Binder::BindMacroSelect(FunctionExpression &function, MacroCatalogEntry *macro_func,
+                                              idx_t depth) {
+
+	auto node = macro_func->function->query_node->Copy();
+
+	D_ASSERT(node->type == QueryNodeType::SELECT_NODE);
+
+	auto &select_node = (SelectNode &)*node;
+	MacroBinding *macro_binding;
+	auto &macro_def = *macro_func->function;
+
+	// validate the arguments and separate positional and default arguments
+	vector<unique_ptr<ParsedExpression>> positionals;
+	unordered_map<string, unique_ptr<ParsedExpression>> defaults;
+	string error = MacroFunction::ValidateArguments(*macro_func, function, positionals, defaults);
+	if (!error.empty()) {
+		// cannot use error below as binder rnot in scope
+		// return BindResult(binder. FormatError(*expr->get(), error));
+		throw BinderException(FormatError(function, error));
+	}
+
+	// create a MacroBinding to bind this macro's parameters to its arguments
+	vector<LogicalType> types;
+	vector<string> names;
+	// positional parameters
+	for (idx_t i = 0; i < macro_def.parameters.size(); i++) {
+		types.emplace_back(LogicalType::SQLNULL);
+		auto &param = (ColumnRefExpression &)*macro_def.parameters[i];
+		names.push_back(param.GetColumnName());
+	}
+	// default parameters
+	for (auto it = macro_def.default_parameters.begin(); it != macro_def.default_parameters.end(); it++) {
+		types.emplace_back(LogicalType::SQLNULL);
+		names.push_back(it->first);
+		// now push the defaults into the positionals
+		positionals.push_back(move(defaults[it->first]));
+	}
+	auto new_macro_binding = make_unique<MacroBinding>(types, names, macro_func->name);
+	new_macro_binding->arguments = move(positionals);
+	macro_binding = new_macro_binding.get();
+
+	// select clause
+	for (auto &select_element : select_node.select_list) {
```

Comment:
You should use `ParsedExpressionIterator::EnumerateQueryNodeChildren` rather than enumerating over all the children manually. The code here is quite complex and there are a number of mistakes here (e.g. the FROM clause can be a subquery, which can hold expressions again, which is not handled here).